### DISCUSSION TITLE
Questions about `jax.lax.cond` in jitted function causing cuStreamSynchronize during dispatch

### DISCUSSION THREAD

**Discussion by skirsten** [jax-ml/jax#13586] [Q&A] [answered]
So this took me forever to track down...
I noticed a performance issue that only occurred in some variants of my code and manifested itself in the dispatcher not feeding the GPU enough work:
![image](https://user-images.githubusercontent.com/1972314/206781341-a2c21ad1-c277-481f-a821-2cb6125a3753.png)
Zoomed in into *focus area*:
![image](https://user-images.githubusercontent.com/1972314/206781747-b89cfcd9-b948-4aaf-a125-a08d6ded472f.png)

As can be seen there is a giant async `MemcpyD2H` waiting and copying a single int back to the host and a `cuStreamSynchronize`.
Only after increasing the number of `jax.lax.cond` calls and noticing that the number of `cuStreamSynchronize` increases as well made I the connection. So it seems that the host is waiting for the result of the condition to conditionally schedule the next kernels.

Now the questions I have is:

1. Why is this copy + sync needed (is my assumption above correct)? I could not find any reference in the documentation that this is blocking...
  Should the [docu](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.cond.html) be updated or a note added?:
```python
def cond(pred, true_fun, false_fun, *operands):
  pred.block_until_ready() # even in compiled functions
  if pred:
    return true_fun(*operands)
  else:
    return false_fun(*operands)

```
2. How can I work around this? It should be beneficially for me to simply compute all possible paths and then use `jnp.where` to select the correct output value. Is there a native way to do that or should I write a simple translation function for `jax.lax.cond`?

Any help is appreciated! :heart: JAX

**Comment by jakevdp** [comment, marked as answer]
Hi, thanks for the question. This is part of the semantics of [`lax.cond`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.cond.html), namely that only a single branch will be evaluated given the condition of the predicate. These semantics can be useful, for example, when you're choosing between two alternatives that may be expensive to compute.

That said, it does have the downside you noted, that it can cause bottlenecks in the computation if `pred` is expensive to compute. If this is the case, you can use [`lax.select`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.select.html) instead, which acts similarly to `lax.cond` but evaluates both branches and so avoids a bottleneck for expensive `pred` inputs.

Note that `jnp.where(pred, a, b)` will lower to `lax.select`.

**Reply by jakevdp** [reply]
I've added some documentation clarifications in https://github.com/google/jax/pull/13589

**Reply by skirsten** [reply]
Awesome, thank you very much! I was able to replace the functions and the performance increased as expected.
